File: VB/Customfigs_VB/VectorDrawCustomFigs_VB/VectorDrawMultiLine.vb
Imports System
Imports System.Runtime.InteropServices
Imports System.ComponentModel
Imports System.Windows.Forms
Imports System.Drawing
Imports VectorDraw.Serialize
Imports VectorDraw.Professional
Imports VectorDraw.Professional.vdPrimaries
Imports VectorDraw.Professional.vdFigures
Imports VectorDraw.Professional.vdCollections
Imports VectorDraw.Professional.vdObjects
Imports VectorDraw.Professional.Constants
Imports VectorDraw.Geometry
Imports VectorDraw.Render
Imports VectorDraw.Professional.Actions
Imports VectorDraw.Actions

Imports VectorDraw.Generics



Public Class VectorDrawOffsetLine
    Inherits vdPolyline
    Implements IvdProxyFigure

#Region "ctor"
    Public Sub VectorDrawOffsetLine()

    End Sub
#End Region

#Region "private properties"
    Private mOffsetcurves As vdArray(Of vdCurves) = Nothing

    Private mLinesColor As vdColor = Nothing
    Private mLinesDistance As Double = 0.3
    Private mNumOfLines As Integer = 2
#End Region

#Region "Public properties"
    'You should use the AddHistory command in your properties so any change will be writen to the history
    'The string in the AddHistory command is the name of the property exactly and should be called before the property change.
    Public Property NumOfLines() As Integer
        Get
            Return mNumOfLines
        End Get
        Set(ByVal Value As Integer)
            AddHistory("NumOfLines", Value)
            mNumOfLines = Value
        End Set
    End Property
    Public Property LinesColor() As vdColor
        Get
            If (mLinesColor Is Nothing) Then mLinesColor = New vdColor(System.Drawing.Color.Red)
            Return mLinesColor
        End Get
        Set(ByVal Value As vdColor)
            AddHistory("LinesColor", Value)
            If (mLinesColor Is Nothing) Then mLinesColor = New vdColor(System.Drawing.Color.Red)
            mLinesColor.CopyFrom(Value)
        End Set
    End Property
    Public Property LinesDistance() As Double
        Get

            Return mLinesDistance
        End Get
        Set(ByVal Value As Double)
            AddHistory("LinesDistance", Value)
            mLinesDistance = Value
        End Set
    End Property
#End Region

#Region "Polyline overrides"
    <Browsable(False)> _
    Public Shadows ReadOnly Property XProperties() As vdXProperties
        Get '{ get { return base.XProperties; } }
            Return MyBase.XProperties
        End Get
    End Property

    <Browsable(False)> _
    Public Overrides Property SPlineFlag() As VdConstSplineFlag
        Get
            Return VdConstSplineFlag.SFlagSTANDARD
        End Get
        Set(ByVal Value As VdConstSplineFlag)
            MyBase.SPlineFlag = VdConstSplineFlag.SFlagSTANDARD
        End Set
    End Property

    Public Overloads Overrides Function Explode() As vdEntities

        Dim Entities As New vdEntities()
        Entities.SetUnRegisterDocument(Document)
        If Document IsNot Nothing Then
            Document.UndoHistory.PushEnable(False)
        End If

        Dim curvesArray As vdArray(Of vdCurves) = Offsetcurves()
        For Each curves As vdCurves In curvesArray
            For Each curve As vdCurve In curves
                curve.PenColor = mLinesColor
                Entities.AddItem(TryCast(curve.Clone(Document), vdFigure))
            Next
        Next
        Dim BaseEnts As vdEntities = MyBase.Explode()
        For Each var As vdFigure In BaseEnts
            Entities.AddItem(TryCast(var.Clone(Document), vdFigure))
        Next


        If Document IsNot Nothing Then
            Document.UndoHistory.PopEnable()
        End If
        Return Entities
    End Function


    Public Overrides Sub Update()
        MyBase.Update()
        Me.mOffsetcurves = Nothing
    End Sub
    Private Function Offsetcurves() As vdArray(Of vdCurves)
        If (Me.mOffsetcurves Is Nothing) Then
            Me.mOffsetcurves = New vdArray(Of vdCurves)
            Dim i As Integer
            For i = 0 To Me.mNumOfLines - 1
                Dim curves As vdCurves = Me.getOffsetCurve((Me.mLinesDistance * (i + 1)))
                If ((Not curves Is Nothing) AndAlso (curves.Count > 0)) Then
                    Me.mOffsetcurves.AddItem(curves)
                End If
            Next i
        End If
        Return Me.mOffsetcurves
    End Function
    Public Overrides Sub InitializeProperties()
        MyBase.InitializeProperties()
        mLinesColor = New vdColor()
        mLinesColor.SetUnRegisterDocument(Me.Document)
        mLinesColor.SystemColor = System.Drawing.Color.Red
        mLinesDistance = 0.3
        mNumOfLines = 2
    End Sub
    'Note that not all values/objects can be saved/serialize by VDF, and you need to write code for such properties in serialize/deserialize override of this object. 
    'You can set value of custom property to any of the following type values: 
    '     String, Double, Int32, Boolean, Byte, Int16, Int64, UInt16, UInt64, UInt32, Single, ushort, System.Color, VectorDraw.Serialize.IVDHandle and IVDSerialise
    Public Overrides Sub Serialize(ByVal serializer As Serializer)
        MyBase.Serialize(serializer)
        serializer.Serialize(mLinesColor, "LinesColor")
        serializer.Serialize(mLinesDistance, "LinesDistance")
        serializer.Serialize(mNumOfLines, "NumOfLines")
    End Sub
    Public Overrides Function DeSerialize(ByVal deserializer As DeSerializer, ByVal fieldname As String, ByVal value As Object) As Boolean
        If MyBase.DeSerialize(deserializer, fieldname, value) Then
            Return True
        ElseIf fieldname = "LinesColor" Then
            mLinesColor = CType(value, vdColor)
        ElseIf fieldname = "LinesDistance" Then
            mLinesDistance = CType(value, Double)
        ElseIf fieldname = "NumOfLines" Then
            mNumOfLines = CType(value, Integer)
        Else : Return False
        End If
        Return True
    End Function
    Public Overrides Sub MatchProperties(ByVal _from As vdPrimary, ByVal thisdocument As vdDocument)
        MyBase.MatchProperties(_from, thisdocument)
        Dim from As VectorDrawOffsetLine = CType(_from, VectorDrawOffsetLine)
        If from Is Nothing Then
            Return
        End If
        mLinesColor = New vdColor()
        mLinesColor.CopyFrom(from.LinesColor)
        mLinesDistance = from.LinesDistance
        mNumOfLines = from.NumOfLines
    End Sub
    Public Overrides Function Draw(ByVal render As vdRender) As vdRender.DrawStatus
        Dim doDraw As vdRender.DrawStatus = MyBase.Draw(render)
        If doDraw = vdRender.DrawStatus.Successed Then
            Dim style As vdGdiPenStyle = New vdGdiPenStyle()
            style.CopyFrom(render.PenStyle)
            style.color = mLinesColor.SystemColor
            render.PushPenstyle(style)

            Dim curvesArray As vdArray(Of vdCurves) = Me.Offsetcurves
            Dim curves As vdCurves
            For Each curves In curvesArray
                Dim curve As vdCurve
                For Each curve In curves
                    Dim pts As gPoints = curve.GetSamplePoints(0, 0)
                    If (pts.Count > 0) Then
                        render.DrawPLine(Me, pts)
                    End If
                Next
            Next

            render.PopPenstyle()
        End If
        Return doDraw
    End Function
    Public Overrides ReadOnly Property BoundingBox() As Box
        Get
            If Not mBoundBox.IsEmpty Then
                Return mBoundBox
            End If
            mBoundBox = New Box()
            Dim var As Vertex
            For Each var In Me.VertexList
                Dim pt As gPoint = CType(var, gPoint)
                mBoundBox.AddPoint(pt)
            Next
            'Make the last offset commited and include the result to the bounding box.
            Dim curvesArray As vdArray(Of vdCurves) = Me.Offsetcurves
            Dim curves As vdCurves
            For Each curves In curvesArray
                Dim curve As vdCurve
                For Each curve In curves
                    Dim pts As gPoints = curve.GetSamplePoints(0, 0)
                    If (pts.Count > 0) Then
                        MyBase.mBoundBox.AddPoints(pts)
                    End If
                Next
            Next


            Return mBoundBox
        End Get
    End Property

    Public Overrides Function IsTableObjectDependOn(ByVal table As VectorDraw.Professional.vdObjects.vdPrimary) As Boolean
        If Deleted = True Then Return False ''always check if this object is deleted
        If MyBase.IsTableObjectDependOn(table) = True Then Return True ''always check the base implamentation
        ''check properties of object that reference the table object.
        If Not LinesColor Is Nothing And LinesColor.IsTableObjectDependOn(table) = True Then Return True
        Return False
    End Function
    Public Overloads Overrides Sub GetTableDependecies(ByVal args As vdTableDependeciesArgs)
        MyBase.GetTableDependecies(args)
        If args.IsObjectFound Then
            Exit Sub
        End If
        If Deleted Then
            Exit Sub
        End If

        If mLinesColor IsNot Nothing Then
            mLinesColor.GetTableDependecies(args)
        End If
    End Sub

#End Region

#Region "Command"
    Public Shared Sub CmdOffsetLine(ByVal document As vdDocument)
        document.Prompt("First Point:")
        Dim ret As Object = document.ActionUtility.getUserPoint
        document.Prompt(Nothing)
        If ((Not ret Is Nothing) AndAlso TypeOf ret Is gPoint) Then
            Dim SPT As gPoint = TryCast(ret, gPoint)
            Dim vertexes As New Vertexes
            vertexes.Add(SPT)
            Dim code As StatusCode = StatusCode.Success
            Do While (code = StatusCode.Success)
                document.Prompt("Next Point:")
                Dim aFig As New ActionOffsetLine(vertexes, document.ActiveLayOut)
                document.ActionAdd(aFig)
                code = aFig.WaitToFinish
                document.Prompt(Nothing)
                If (code = StatusCode.Success) Then
                    vertexes = DirectCast(aFig.Entity, VectorDrawOffsetLine).VertexList
                End If
            Loop
            Dim line As New VectorDrawOffsetLine
            document.UndoHistory.PushEnable(False)
            line.SetUnRegisterDocument(document)
            line.InitializeProperties()
            line.setDocumentDefaults()
            line.VertexList = vertexes
            line.Transformby(document.User2WorldMatrix)
            document.UndoHistory.PopEnable()
            document.ActionLayout.Entities.AddItem(DirectCast(line, vdFigure))
            document.ActionDrawFigure(line)
        End If
    End Sub


#End Region
End Class

Public Class ActionOffsetLine
    Inherits ActionEntity
    Private line As VectorDrawOffsetLine

    Public Sub New(ByVal vertexes As Vertexes, ByVal layout As vdLayout)
        MyBase.New(layout.Document.User2WorldMatrix.Transform(vertexes.Last()), layout)
        line = New VectorDrawOffsetLine

        line.SetUnRegisterDocument(layout.Document)
        line.InitializeProperties()

        line.setDocumentDefaults()
        vertexes.Add(vertexes.Last())
        line.VertexList = vertexes
    End Sub

    Public Overrides ReadOnly Property HideRubberLine() As Boolean
        Get
            Return True
        End Get
    End Property
    Public Overrides ReadOnly Property Entity() As vdFigure
        Get
            Return line
        End Get
    End Property
    Protected Overrides Sub OnMyPositionChanged(ByVal NewPosition As gPoint)
        line.VertexList.Last().CopyFrom(NewPosition)
    End Sub
    Public Overrides ReadOnly Property needUpdate() As Boolean
        Get
            Return True
        End Get
    End Property

End Class

File: CustomObjects/vdPolygon/VectorDrawMultiLine.cs
using System;
using System.Runtime.InteropServices;
using System.ComponentModel;
using System.Windows.Forms;
using VectorDraw.Serialize;
using VectorDraw.Professional;
using VectorDraw.Professional.vdPrimaries;
using VectorDraw.Professional.vdFigures;
using VectorDraw.Professional.vdCollections;
using VectorDraw.Professional.vdObjects;
using VectorDraw.Professional.Constants;
using VectorDraw.Geometry;
using VectorDraw.Render;
using VectorDraw.Professional.Actions;
using VectorDraw.Actions;
using VectorDraw.Generics;

//This custom object also explains how to export and import via xProperties the properties of the object into a dwg file.
//Also check the code to the CustomObjects project at the Form1.cs

namespace VectorDrawCustomFigs
{
    public class VectorDrawOffsetLine : vdPolyline, IvdProxyFigure
    {
        #region ctor
        public VectorDrawOffsetLine()
        {

        }
        public VectorDrawOffsetLine(vdDocument doc)
            : this()
        {
            SetUnRegisterDocument(doc);
            setDocumentDefaults();
        }
        #endregion

        #region private properties
        private vdArray<vdCurves> mOffsetcurves = null;
        private vdColor mLinesColor = null;
        private double mLinesDistance = 0.3;
        private int mNumOfLines = 2;
        #endregion

        #region public properties
        //You should use the AddHistory command in your properties so any change will be writen to the history
        //The string in the AddHistory command is the name of the property exactly and should be called before the property change.
        public int NumOfLines
        {
            get
            {
                return mNumOfLines;
            }
            set
            {
                AddHistory("NumOfLines", value);
                mNumOfLines = value;
            }
        }
        public vdColor LinesColor
        {
            get
            {
                if (mLinesColor == null) mLinesColor = new vdColor(System.Drawing.Color.Red);
                return mLinesColor;
            }
            set
            {
                AddHistory("LinesColor", value);
                if (mLinesColor == null) mLinesColor = new vdColor(System.Drawing.Color.Red);
                mLinesColor.CopyFrom(value);
            }
        }
        public double LinesDistance
        {
            get
            {
                return mLinesDistance;
            }
            set
            {
                AddHistory("LinesDistance", value);
                mLinesDistance = value;
            }
        }
        #endregion

        #region Polyline overrides
        //This code is to hide the Xproperties property of this custom object for the properties list.
        [Browsable(false)]
        public new vdXProperties XProperties { get { return base.XProperties; } } 
    
        [Browsable(false)]
        public override VdConstSplineFlag SPlineFlag
        {//Do not implement the SPline Flag for this object.
            get
            {
                return VdConstSplineFlag.SFlagSTANDARD;
            }
            set
            {
                base.SPlineFlag = VdConstSplineFlag.SFlagSTANDARD;
            }
        }
        public override bool Break(gPoint p1, gPoint p2, out vdFigure newEntity)
        {
            return base.Break(p1, p2, out newEntity);
        }

        public override vdEntities Explode()
        {
            // In order to save this object to a DWG/DXF file and when reopened in this application to maintain and recalculate the custom object we do the following
            // Whatever information you want to preserve to the exported file it must be added as xProperty to the exploded figures , since these entities will be saved.
            // We also Add a string xproperty to all the exported figures in order to identify them while opening the file.
            // The properties of the custom object are added to the first exploded figure.
            // Note that since when openoing the file a new object will be created the Handle of the custom object will not be preserved.
            // The rest of the code is in the void ActiveDocument_OnAfterOpenDocument(object sender) event of the CustomObjects project (Form1.cs).
            
            vdEntities Entities = new vdEntities();
            Entities.SetUnRegisterDocument(Document);
            if (Document != null) Document.UndoHistory.PushEnable(false);

            vdArray<vdCurves> curvesArray = Offsetcurves();
            foreach (vdCurves curves in curvesArray)
            {
                foreach (vdCurve curve in curves)
                {
                    curve.PenColor = mLinesColor;
                    vdXProperty xprop = curve.XProperties.FindName("vdDWGXprop");
                    if (xprop == null)
                    {
                        vdXProperty xprop1 = curve.XProperties.Add("vdDWGXprop");
                        xprop1.PropValue = "vdDWGXprop";
                    }
                    else 
                    {
                        xprop.PropValue = "vdDWGXprop";
                    }
                    Entities.AddItem(curve.Clone(Document) as vdFigure);
                }
            }

            vdEntities BaseEnts = base.Explode();
            foreach (vdFigure  var in BaseEnts)
            {
                vdFigure fig = var.Clone(Document) as vdFigure;
                vdXProperty xprop = fig.XProperties.FindName("vdDWGXprop");
                if (xprop == null)
                {
                    vdXProperty xprop1 = fig.XProperties.Add("vdDWGXprop");
                    xprop1.PropValue = "vdDWGXprop";
                }
                else 
                {
                    xprop.PropValue = "vdDWGXprop";
                }

                Entities.AddItem(fig);
            }

            if (Entities.Count > 0)
            {
                vdFigure fig = Entities[0];
                vdXProperty xprop = fig.XProperties.FindName("NumLines");
                xprop.PropValue = this.NumOfLines;

                xprop = fig.XProperties.FindName("LinesColor");
                xprop.PropValue = this.LinesColor.ToInt32();

                xprop = fig.XProperties.FindName("LinesDistance");
                xprop.PropValue = this.LinesDistance;

                xprop = fig.XProperties.FindName("PenColor");
                xprop.PropValue = this.PenColor.ToInt32();

                xprop = fig.XProperties.Add("NumberOfExplodedEntities");
                xprop.PropValue = 7;

                vdEntities ents = this.Owner as vdEntities;
                if (ents != null)
                {
                    xprop = fig.XProperties.Add("vdRealPosition");
                    int  pos = ents.GetObjectRealPosition(this);
                    xprop.PropValue = pos;
                }
            }


            if (Document != null) Document.UndoHistory.PopEnable();
            return Entities;
        }
        public override void Update()
        {
            base.Update();
            mOffsetcurves = null;
        }
        private vdArray<vdCurves> Offsetcurves()
        {
            if (mOffsetcurves != null) return mOffsetcurves;
            Vector V = null;
            Matrix matt = new Matrix();
            vdPolyline tmppl = this.Clone(Document) as vdPolyline;
            if (this.VertexList.Count > 2)
            {
                V = tmppl.VertexList.GetNormal();
                matt = new Matrix();
                if (V != null)
                {
                    matt.ApplyWCS2ECS(V);
                    tmppl.Transformby(matt);
                }
            }
            mOffsetcurves = new vdArray<vdCurves>();
            for (int i = 0; i < mNumOfLines; i++)
            {
                vdCurves curves = tmppl.getOffsetCurve(mLinesDistance * (i + 1));
                if (curves != null && curves.Count > 0 && mOffsetcurves != null)
                {
                    //foreach (vdCurve var in curves) var.Transformby(matt.GetInvertion());
                    mOffsetcurves.AddItem(curves);
                }
            }
            return mOffsetcurves;
        }
        protected override void OnDocumentSelected(vdDocument document)
        {
            LinesColor.SetUnRegisterDocument(document);

            if (this.XProperties.Count == 0)
            {
                this.XProperties.Add("vdDWGXprop");
                vdXProperty xprop = this.XProperties.Add("NumLines");
                xprop.PropValue = this.NumOfLines;

                xprop = this.XProperties.Add("LinesColor");
                xprop.PropValue = this.LinesColor.ToInt32();

                xprop = this.XProperties.Add("LinesDistance");
                xprop.PropValue = this.LinesDistance;

                xprop = this.XProperties.Add("PenColor");
                xprop.PropValue = this.PenColor.ToInt32();
            }
        }
        public override void InitializeProperties()
        {
            base.InitializeProperties();
            mLinesColor = new vdColor();
            mLinesColor.SetUnRegisterDocument(this.Document);
            mLinesColor.SystemColor = System.Drawing.Color.Red;
            mLinesDistance = 0.3;
            mNumOfLines = 2;


            
        }

        //Note that not all values/objects can be saved/serialize by VDF, and you need to write code for such properties in serialize/deserialize override of this object. 
        //You can set value of custom property to any of the following type values: 
        //     String, Double, Int32, Boolean, Byte, Int16, Int64, UInt16, UInt64, UInt32, Single, ushort, System.Color, VectorDraw.Serialize.IVDHandle and IVDSerialise
        public override void Serialize(Serializer serializer)
        {
            base.Serialize(serializer);
            serializer.Serialize(mLinesColor, "LinesColor");
            serializer.Serialize(mLinesDistance, "LinesDistance");
            serializer.Serialize(mNumOfLines, "NumOfLines");
        }
        public override bool DeSerialize(DeSerializer deserializer, string fieldname, object value)
        {
            if (base.DeSerialize(deserializer, fieldname, value)) return true;
            else if (fieldname == "LinesColor") mLinesColor = (vdColor)value;
            else if (fieldname == "LinesDistance") mLinesDistance = (double)value;
            else if (fieldname == "NumOfLines") mNumOfLines = (int)value;

            else return false;
            return true;
        }
        public override void MatchProperties(vdPrimary _from, vdDocument thisdocument)
        {
            base.MatchProperties(_from, thisdocument);
            VectorDrawOffsetLine from = _from as VectorDrawOffsetLine;
            if (from == null) return;
            mLinesColor = new vdColor();
            mLinesColor.CopyFrom(from.LinesColor);
            mLinesDistance = from.LinesDistance;
            mNumOfLines = from.NumOfLines;
        }
        public override vdRender.DrawStatus Draw(vdRender render)
        {
            if (ListStart(render)) return render.StatusDraw;
            vdRender.DrawStatus doDraw = base.Draw(render);
            if (doDraw == vdRender.DrawStatus.Successed)
            {
                render.PushDrawFigureList(this);

                vdGdiPenStyle style = new vdGdiPenStyle();
                style.CopyFrom(render.PenStyle);
                style.color = mLinesColor.SystemColor;
                render.PushPenstyle(style);

                vdArray<vdCurves> curvesArray = Offsetcurves();
                if (curvesArray == null) return vdRender.DrawStatus.Successed;
                foreach (vdCurves curves in curvesArray)
                {
                    foreach (vdCurve curve in curves)
                    {
                        gPoints pts = curve.GetSamplePoints(0, 0);
                        ECSMatrix.Transform(pts);
                        if (pts.Count > 0) render.DrawPLine(this, pts,base.Thickness);
                    }
                }
                render.PopPenstyle();

                render.PopDrawFigureList();
            }
            if (ListEnd(render)) doDraw = render.StatusDraw;
            return doDraw;
        }
        public override Box BoundingBox
        {
            get
            {
                if (!mBoundBox.IsEmpty) return mBoundBox;
                mBoundBox = new Box();
                foreach (Vertex var in this.VertexList)
                {
                    gPoint pt = var as gPoint;
                    mBoundBox.AddPoint(pt);
                }
                //Make the last offset commited and include the result to the bounding box.
                vdArray<vdCurves> curvesArray = Offsetcurves();
                if (curvesArray == null) return mBoundBox;
                foreach (vdCurves curves in curvesArray)
                {
                    foreach (vdCurve curve in curves)
                    {
                        gPoints pts = curve.GetSamplePoints(0, 0);
                        if (pts.Count > 0) mBoundBox.AddPoints(pts);
                    }
                }

                return mBoundBox;
            }
        }
        public override bool IsTableObjectDependOn(vdPrimary table)
        {
            if (Deleted) return false;//always check if this object is deleted
            if (base.IsTableObjectDependOn(table)) return true;//always check the base implamentation
            //check properties of object that reference the table object.
            if (LinesColor != null && LinesColor.IsTableObjectDependOn(table)) return true;
            return false;
        }
        public override void GetTableDependecies(vdTableDependeciesArgs args)
        {
            base.GetTableDependecies(args);
            if (args.IsObjectFound) return;
            if (Deleted) return;

            if (LinesColor != null) LinesColor.GetTableDependecies(args);
        }
        #endregion

        #region Command
        public static void CmdOffsetLine(vdDocument document)
        {
            gPoint SPT;
            document.Prompt("First Point:");
            object ret = document.ActionUtility.getUserPoint();
            document.Prompt(null);
            if (ret == null || !(ret is gPoint)) return;
            SPT = ret as gPoint;
            Vertexes vertexes = new Vertexes();
            vertexes.Add(SPT);
            StatusCode code = StatusCode.Success;
            while (code == StatusCode.Success)
            {
                document.Prompt("Next Point:");
                ActionOffsetLine aFig = new ActionOffsetLine(vertexes, document.ActiveLayOut);
                document.ActionAdd(aFig);
                code = aFig.WaitToFinish();

                document.Prompt(null);
                if (code == StatusCode.Success)
                {
                    vertexes =new Vertexes (((VectorDrawOffsetLine)aFig.Entity).VertexList);
                }
            }
            VectorDrawOffsetLine line = new VectorDrawOffsetLine();
            document.UndoHistory.PushEnable(false);
            line.SetUnRegisterDocument(document);
            line.InitializeProperties();
            line.setDocumentDefaults();
            vertexes.RemoveEqualPoints();
            line.VertexList = vertexes;
            line.Transformby(document.User2WorldMatrix);
            document.UndoHistory.PopEnable();
            document.ActionLayout.Entities.AddItem(line);
            document.ActionDrawFigure(line);
        }
        #endregion
    }

    public class ActionOffsetLine : ActionEntity
    {
        private VectorDrawOffsetLine mline;
        public ActionOffsetLine(Vertexes vertexes, vdLayout layout)
            : base(layout.Document.User2WorldMatrix.Transform(vertexes.Last()), layout)
        {
            mline = new VectorDrawOffsetLine();
            mline.SetUnRegisterDocument(layout.Document);
            mline.InitializeProperties();
            mline.setDocumentDefaults();
            vertexes.Add(vertexes.Last());
            mline.VertexList = vertexes;
        }
        public override bool HideRubberLine
        {
            get
            {
                return true;
            }
        }
        public override vdFigure Entity
        {
            get
            {
                return mline;
            }
        }
        protected override void OnMyPositionChanged(gPoint newPosition)
        {
            mline.VertexList.Last().CopyFrom(newPosition);
        }
        public override bool needUpdate
        {
            get
            {
                return true;
            }
        }
        

    }

}
